plugins/ScheduledReports/SubscriptionModel.php

```diff
@@ -0,0 +1,188 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+namespace Piwik\Plugins\ScheduledReports;
+
+use Piwik\Access;
+use Piwik\API\Request;
+use Piwik\Common;
+use Piwik\Db;
+use Piwik\DbHelper;
+use Piwik\Piwik;
+
+class SubscriptionModel
+{
+    private static $rawPrefix = 'report_subscriptions';
+    private $table;
+
+    public function __construct()
+    {
+        $this->table = Common::prefixTable(self::$rawPrefix);
+    }
+
+    public function unsubscribe($token)
+    {
+        $details = $this->getSubscription($token);
+
+        if (empty($details)) {
+            return false;
+        }
+
+        $email = $details['email'];
+
+        $report = Access::doAsSuperUser(function() use ($details) {
+            $reports = Request::processRequest('ScheduledReports.getReports', array(
+                'idReport'    => $details['idreport'],
+            ));
+            return reset($reports);
+        });
+
+        if (empty($report)) {
```

Comment:
In the case where there's no report for a subscription, should the subscription be removed?

Reply:
Note: think this is an edge case, so could put this in a new ticket if important

Reply:
changed